pyproject.toml

```diff
@@ -155,6 +155,84 @@ version_files = [
     'docs/package.json:^  "version":',
 ]
 
+[tool.mypy]
+files = "src"
+
+# The following whitelist is used to allow for incremental adoption
+# of MyPy. Modules should be removed from this whitelist as and when
+# their respective type errors have been addressed. No new modules
+# should be added to this whitelist.
+[[tool.mypy.overrides]]
+module = [
+  "meltano.api.*",
+  "meltano.cli.*",
+  "meltano.migrations.*",
+  "meltano.oauth.*",
+  # Way too many modules at the root of meltano.core to tackle them all at once
+  # so listing them individually here.
+  "meltano.core.cli_messages",
+  "meltano.core.config_service",
+  "meltano.core.connection_service",
+  "meltano.core.db",
+  "meltano.core.discovery_file",
+  "meltano.core.elt_context",
+  "meltano.core.environment",
+  "meltano.core.environment_service",
+  "meltano.core.error",
+  "meltano.core.extract_utils",
+  "meltano.core.extractor",
+  "meltano.core.meltano_file",
+  "meltano.core.meltano_invoker",
+  "meltano.core.migration_service",
+  "meltano.core.models",
+  "meltano.core.plugin_discovery_service",
+  "meltano.core.plugin_install_service",
+  "meltano.core.plugin_invoker",
+  "meltano.core.plugin_location_remove",
+  "meltano.core.plugin_lock_service",
+  "meltano.core.plugin_remove_service",
+  "meltano.core.plugin_test_service",
+  "meltano.core.project",
+  "meltano.core.project_add_service",
+  "meltano.core.project_files",
+  "meltano.core.project_init_service",
+  "meltano.core.project_plugins_service",
+  "meltano.core.project_settings_service",
+  "meltano.core.schedule",
+  "meltano.core.schedule_service",
+  "meltano.core.select_service",
+  "meltano.core.setting",
+  "meltano.core.setting_definition",
+  "meltano.core.settings_service",
+  "meltano.core.settings_store",
+  "meltano.core.sqlalchemy",
+  "meltano.core.state_service",
+  "meltano.core.task_sets",
+  "meltano.core.task_sets_service",
+  "meltano.core.transform_add_service",
+  "meltano.core.upgrade_service",
+  "meltano.core.validation_service",
+  "meltano.core.venv_service",
+  "meltano.core.yaml",
```

Comment:
This probably calls for some refactoring to put some of these in dedicated submodules.